antispam.py

```diff
@@ -27,30 +29,39 @@ def is_spam(self, identifier):
         """
         logger.debug("Checking for spam for identifier: %s", identifier)
         self._lookup.setdefault(identifier, []).append(time.time())
-        return len(self._lookup[identifier]) > self._count
+        is_spam = len(self._lookup[identifier]) > self._count
+        logger.debug("%s is spam: %s", identifier, is_spam)
+        return is_spam
 
-    def clean(self):
+    @staticmethod
+    def clean():
         """
         cleanup times that are pass the configured timeout
         :return: void
         """
         logger.debug("Cleaning up..")
-        threshold = time.time() - self._timeout
-        for identifier in self._lookup:
-            pruned = [t for t in self._lookup[identifier] if t > threshold]
-            if pruned:
-                self._lookup[identifier] = pruned
-            else:
-                self._lookup.pop(identifier, None)
-
-
-def spam_protect(func):
-    def call(*args, **kwargs):
-        if len(args) is 3:
-            update = args[2]
-            if AntiSpam().is_spam(update.message.from_user.id):
-                update.message.reply_text("You are too talkative, pls wait a bit and try again")
-                return
-        return func(*args, **kwargs)
-    return call
+        for instance in AntiSpam.__refs__:
```

Comment:
this looks kind of dangerous to me
` The original object can be retrieved by calling the reference object if the referent is still alive; if the referent is no longer alive, calling the reference object will cause None to be returned. `
and moreover the object can be destroyed by the GC at any time, how can we guarantee that it's still alive. Documentation says about callback for this case. But still looks UB

Reply:
I see, I assumed there would be some locking involved when iterating.

As a solution:

```
        instanses = set([])
        instanses .update(_dangling)
        for instance in instances ....
```